### PR TITLE
feat(daplink_flash): Add file-IO driver on top of the DaplinkBridge.

### LINKED ISSUE #12: feat(daplink_flash): Implement Arduino driver.

### LINKED ISSUE #32: test(daplink_flash): Add mock unit tests.
## Description

Add desktop unit tests for the daplink_flash Arduino driver using MockWire.

## Tasks

- [ ] Create `test/test_daplink_flash/test_main.cpp`
- [ ] Test device ID / WHO_AM_I register read
- [ ] Test key methods (measurement reads, power on/off, etc.)
- [ ] Test error handling (device not found, timeout)
- [ ] All tests pass with `pio test -e native`

## Reference

- MicroPython mock scenarios: [tests/scenarios/daplink_flash.yaml](https://github.com/steamicc/micropython-steami-lib/blob/main/tests/scenarios/)
- Use the MicroPython scenarios as a guide for what to test

## Depends on

- #20 (test framework)
- #21 (MockWire)
- Driver implementation issue

### LINKED ISSUE #157: fix(hts221): humidity() right after begin() occasionally returns 0 via the auto-trigger path
## Context

While addressing a Copilot finding on PR #153 (level-1 hardware unit test for HTS221), the test suite uncovered an edge case in the driver's auto-trigger path:

A `humidity()` call made right after `begin()` — with no prior `setContinuous(...)` — occasionally returns `0.0f` instead of a real reading. The temperature path on the same trigger comes back correctly.

Worked around in PR #153 by adding `sensor.setContinuous(HTS221_ODR_1_HZ)` inside the test `setUp()`. The test now passes reliably, but the underlying behaviour is in the driver itself, not the test.

## Reproduction

```cpp
HTS221 sensor(internalI2C);

void test_humidity_after_begin_only(void) {
    sensor.begin();
    float h = sensor.humidity();
    // h occasionally observed as 0.0f instead of ~40-60% RH
}
```

Frequency: not 100%, but reproducible enough on real silicon that the plausibility window `[10, 90]` flagged it during the PR #153 hardware run before the workaround was added.

## Hypothesis

`begin()` ends with `powerOff()` (CTRL1.PD = 0). The first `humidity()` call enters `read()`, sees `isPoweredOn() == false`, and routes through `triggerOneShot()` + `waitForDataReady()`.

`waitForDataReady()` polls `STATUS_REG.H_DA & T_DA` with a 100 ms timeout. Most likely the H/T registers are read before the humidity conversion has settled — `T_DA` and `H_DA` may flip true before both registers actually hold the freshly converted values, especially at default oversampling. `computeHumidity()` then receives a raw int16 that resolves to a small negative float, which the existing clamp turns into `0.0f`.

Possible angles to validate / fix:

* Verify with the datasheet whether reading `H_DA | T_DA` is sufficient or whether each channel's `*_DA` should be polled independently before reading its register.
* Consider re-checking `dataReady()` after the register read or adding a short discard-first-sample step inside `read()` when entering from a powered-down state.
* Alternatively, document that callers must call `setContinuous(...)` before `humidity()` and remove the auto-trigger path — but that breaks the existing API contract documented in [`lib/hts221/README.md`](lib/hts221/README.md): "If the part is powered down when a read is requested, the driver auto-triggers a one-shot measurement, polls `dataReady()` with a timeout, and returns the result. The caller doesn't have to manage modes manually."

The third option contradicts the current design principle, so the fix should keep auto-trigger working and make it actually return real data on the first read.

## Acceptance criteria

- [ ] Reproduce the bug deterministically (or characterise its frequency) on a real STeaMi board with a tight loop of `begin()` + `humidity()`.
- [ ] Identify the root cause (timing, register ordering, oversampling default, or other).
- [ ] Fix the auto-trigger path so a `humidity()` call right after `begin()` returns a real value, with the same reliability as the same call after `setContinuous(...)`.
- [ ] Add a native regression test if the failure mode is reproducible against the Wire mock; add a hardware test otherwise.
- [ ] Once the driver is fixed, the `setContinuous(HTS221_ODR_1_HZ)` line in `tests/hardware/test_hts221/setUp()` becomes optional — leaving it in is fine but it should no longer be load-bearing.

## Out of scope

- The temperature path appears unaffected. No change required there unless the investigation reveals the same root cause.
- Other drivers might inherit similar auto-trigger logic eventually; this issue is scoped to HTS221.

### PR DESCRIPTION
**Pull request by nedseb**
## Goal

Add the high-level file-IO driver layered on `DaplinkBridge` (merged in #161). The DAPLink firmware exposes the on-board SPI flash as a FAT-style file system over I²C; this PR wraps that protocol into an Arduino driver matching the MicroPython sister API.

Closes #12.
Closes #32.
Partial #44 — the non-destructive hardware smoke is included; the destructive integration suite is deferred (see [Known limitation](#known-limitation--clearflash-on-real-hardware) below).

## What's in the PR

- **`lib/daplink_flash/`** — the driver itself (`DaplinkFlash` class), register constants, README with destructive-ops guidance, library metadata, keywords.
- **`lib/daplink_bridge/`** — two cross-driver primitives (`sendCommand` + `readResponse`) and a dedicated response-stream register (`REG_RESPONSE = 0x82`) so command writes and response reads use distinct I²C addresses.
- **`tests/native/test_daplink_flash/`** — 19 host-side unit tests against the `TwoWire` mock. Covers happy paths plus regression locks for the three review-found bugs (multi-chunk read past the 0xFF wrap, partial-write offset semantics, `setFilename` bool surface).
- **`tests/hardware/test_daplink_flash/`** — 2 non-destructive smoke tests (`begin()` + printable-ASCII `getFilename()`); runs as part of `make test-hardware` without touching flash content.
- **`src/main.cpp`** — driver wired into the smoke-test sketch alongside the other on-board drivers (per #166).

## Public API (highlights)

| Method | Purpose |
|---|---|
| `bool begin()` | Re-probes the bridge; `false` if absent or WHO_AM_I mismatch. |
| `bool setFilename(name, ext)` | Sets the 8.3 filename; `false` on null arg or bridge error. |
| `FilenameResult getFilename()` | Reads the current 8.3 filename, trimmed. |
| `bool clearFlash()` | **Destructive.** See the limitation below. |
| `size_t write(data, length)` | Append-style write; returns bytes actually written (0, partial, or full). |
| `bool readSector(sector, buf)` | 256-byte sector read. |
| `size_t readUntilSentinel(buf, max)` | Stream bytes until the first 0xFF. |

Every method that talks to the bridge propagates the bridge's busy-timeout and error register: `false` / `0` on failure, deterministic success otherwise. See the README for the full surface and call patterns.

## Known limitation — `clearFlash()` on real hardware

`clearFlash()` is known to brick STeaMi boards into DAPLink maintenance mode. The I²C transaction is byte-identical to the working MicroPython `clear_flash()`, and the F103 firmware is supposed to wipe only the external W25Q64 (not the F103 internal flash that holds the interface firmware), so the bug sits in the firmware itself. Tracked in **[`steamicc/DAPLink#9`](https://github.com/steamicc/DAPLink/issues/9)**.

Until that lands:
- `clearFlash()` stays in the public API (parity with MicroPython); the README opens with a `⚠️ Destructive operations` section and the quick-start does *not* call it.
- The destructive integration suite was reverted (initially added in `2bb70f5`, removed in `3f522e2`); shipping it would mean every `make test-integration` wipes the interface firmware. Re-add once DAPLink#9 lands.
- The non-destructive hardware smoke stays in place.

## Follow-ups

- **#177** — rename `daplink_flash.{h,cpp,_const.h}` → `DaplinkFlash.{h,cpp,_const.h}` (PascalCase, matching `DaplinkBridge.h`, `HTS221.h`, etc.). Deferred to keep this PR focused on runtime fixes.
- **#176** — broaden clang-tidy coverage to catch the class of bugs surfaced by this review wave.

## Test plan

- [x] `make ci` (lint + format-check + native) — **101/101 native PASS**
- [x] `make test-hardware` on a wired STeaMi — daplink_flash smoke PASS
- [x] `make build` (steami env) — smoke sketch links and runs `daplink_flash.begin()` on the board
- [ ] `make test-integration` — deferred to after DAPLink#9

### CODE REVIEW

**Comment by nedseb** [comment]
Aline, 
j'ai ouvert cette PR en draft pour qu'on puisse parler du contenu de ta branche `feat/daplink-flash-driver` avant que tu finalises. Je l'ai rebasée sur main (qui inclut maintenant #161 — ma version réécrite de ton driver bridge), puis j'ai poussé deux commits pour la débloquer côté compilation. Voici une première revue pour t'orienter sur ce qui reste.

## Ce qui se passe pendant le rebase

J'ai droppé tes 5 commits sur `daplink_bridge` (driver, tests, README, fixes) parce qu'ils sont déjà sur main sous une forme différente après #161 — sinon git aurait recréé ta version qui entrait en conflit avec celle de main. Ce qui reste, c'est ton travail pur sur `daplink_flash` :

- `feat(daplink_flash): Add DaplinkFlash driver.`
- `feat(daplink_flash): Add DaplinkFlash native unit tests.`
- `docs(daplink_flash): Add DaplinkFlash README.`
- `feat(daplink_flash): Add library.properties files.`
- `fix(daplink_flash): Declare daplink_bridge as a library dependency.`

C'est sur quoi cette revue porte.

## J'ai déjà adapté le code pour qu'il compile contre main

Toutes tes méthodes `DaplinkFlash` appelaient le bridge avec **l'ancienne API** (la tienne d'avant #161). Sur main, le bridge a maintenant une `DaplinkBridge` (PascalCase, refactorée) avec une surface publique différente, et toutes les primitives bas-niveau (`writeFrame`, `readBlock`, `error`, `waitNotBusy`) sont privées.

Plutôt que d'attendre que tu te débloques toute seule, j'ai poussé deux commits sur la branche :

- `a12c090 feat(daplink_bridge): Add sendCommand/readResponse for sibling drivers.` — expose une API "frame-level" générique sur le bridge, conçue pour que n'importe quel sibling (flash, futur driver) puisse composer son protocole par-dessus :

  ```cpp
  bool sendCommand(uint8_t cmd, const uint8_t* payload = nullptr, size_t payloadLen = 0,
                   uint32_t timeoutMs = DAPLINK_BRIDGE_WRITE_TIMEOUT_MS);
  size_t readResponse(uint8_t cmd, uint8_t* buf, size_t maxLen,
                      uint32_t timeoutMs = DAPLINK_BRIDGE_READ_TIMEOUT_MS);
  ```

  `sendCommand` frame `[CMD | payload...]`, attend que le bridge ne soit plus busy, et retourne `false` si le device répond une erreur. `readResponse` fait le write-then-read équivalent et chunke en interne sous `MAX_READ_CHUNK`. Les méthodes existantes (`clearConfig` / `writeConfig` / `readConfig`) restent comme convenience pour la zone de config.

- `55eda25 fix(daplink_flash): Adapt to the new DaplinkBridge public API.` — migre ton driver vers la nouvelle API. Concrètement :

  | Tu avais | C'est devenu |
  |---|---|
  | `class daplink_bridge` (lowercase) | `class DaplinkBridge` |
  | `#include "daplink_bridge.h"` | `#include "DaplinkBridge.h"` |
  | `_bridge->writeTo(buf, len)` + `error()` check | `_bridge->sendCommand(cmd, payload, len)` |
  | `_bridge->readBlock(reg, buf, len)` | `_bridge->readResponse(cmd, buf, len)` |
  | `_bridge->writeCmd(cmd)` + `waitBusy()` | `_bridge->sendCommand(cmd)` |
  | `library.properties` `depends=STeaMi daplink bridge` | `depends=STeaMi DAPLink Bridge` (matche le `name=` du bridge) |

  Les tests natifs ont été mis à jour pour utiliser les noms `DAPLINK_BRIDGE_REG_*` / `DAPLINK_BRIDGE_STATUS_*` au lieu des noms non-préfixés.

**65/65 tests natifs passent maintenant**, mais j'ai dû désactiver 2 tests avec un TODO explicite (`test_read_stops_at_sentinel`, `test_read_limited_by_maxlen`) — ils segfaultent à cause d'un bug protocolaire dans `readSector` que je détaille plus bas dans Design.

À toi de prendre la suite à partir de là — il reste les conventions, le design API et le rework des tests.

## Conventions à appliquer (familier après #143 / #152 / #161)

Les mêmes points que sur tes PR précédentes. Liste à cocher au passage :

- **`begin()` manque sur `DaplinkFlash`.** L'utilisateur n'a aucun moyen de savoir si le bridge est joignable avant de tenter `clearFlash()` (qui prend ~secondes à timer out s'il n'est pas là).
- **Constants sans préfixe** dans `daplink_flash_const.h` (`CMD_WRITE_DATA`, `MAX_SECTORS`, etc.). Avec le bridge sur main qui a maintenant `DAPLINK_BRIDGE_CMD_*`, il y a un risque de collision de namespace global si tu importes les deux. À préfixer en `DAPLINK_FLASH_CMD_*`.
- **README ne mentionne pas `begin()`** et la quick-start utilise encore `daplink_bridge` lowercase dans les exemples (j'ai corrigé le code mais pas la doc).

## Design API à challenger

Quelques points où le shape de l'API se discute :

- **Bug protocolaire dans `readSector`** : la méthode prend un `sectorIndex` mais ne le passe jamais au bridge. Mes deux tests désactivés taperaient ce path. Il faut soit ajouter un `seekSector(uint16_t index)` côté firmware DAPLink (et l'envoyer avant le `readResponse`), soit embarquer l'index dans la commande `CMD_READ_SECTOR`. À trancher avec le firmware DAPLink.
- **`size_t read(uint8_t* result, size_t maxLen, bool limitLen = false)`** — le `limitLen` change la sémantique du même appel (avec `false`, on lit jusqu'au sentinel `0xFF` ; avec `true`, on lit jusqu'à `maxLen`). Deux méthodes distinctes seraient plus claires : `readUntilSentinel(buf, maxLen)` et `readN(buf, n)`. Pas un blocker, mais plus prévisible.
- **`size_t write(...)` retourne 0 sur erreur** — mais aussi 0 si `length == 0` est passé. Ambiguïté. Pattern HTS221 : signaler l'erreur via une autre voie. Ici je verrais bien `bool write(data, len, size_t* bytesWritten = nullptr)`.
- **`writeLine` ne propage pas l'échec de `write`** — il retourne toujours `len + 1` même si le `write` sous-jacent a renvoyé 0. À corriger.
- **`uint8_t buf[len + 1]` dans `writeLine`** — VLA non-standard en C++. Préférer un `std::vector<uint8_t>` ou un buffer fixe avec borne.

## Tests — les mêmes patterns à corriger qu'avant

Tu as la même structure de test que sur le bridge avant qu'on la corrige. Notamment :

- `test_write_sends_data` : `memcmp(&w.value, data, strlen(data))` est encore l'UB qu'on a vue dans #161. Le test passe en fait sans rien valider.
- `test_clear_flash_sends_cmd` : `TEST_ASSERT_TRUE(true)` — n'assert rien d'utile.
- `test_read_sector_sends_correct_command` : pareil, `TEST_ASSERT_TRUE(true)`.
- Pas de test `begin()` (puisque pas de `begin()`).
- Pas de tests sur les paths d'erreur du bridge (busy timeout, error register != 0 quand on écrit).
- Les 2 tests désactivés (`test_read_stops_at_sentinel`, `test_read_limited_by_maxlen`) à réactiver une fois le bug `readSector` corrigé.

Le fix #161 te montre le pattern propre — une fois que tu fixes les autres points, les tests pourront utiliser `check_begin` / `check_who_am_i` du shared.

## Manques pour que ce soit complet

- **Section `DaplinkFlash` dans `src/main.cpp`** : convention #166 demande qu'un nouveau driver implémenté soit câblé au smoke-test. Comme le driver est destructif (`clearFlash`), peut-être juste un `flash.begin()` qui lit le filename actuel sans rien écrire.
- **`writeLine` documenté nulle part**.
- **Méthode `getFilename()` qui retourne par valeur un `FilenameResult`** — OK pour cette taille (12 bytes), mais à confirmer côté API si tu veux aussi `getFilename(char* nameOut, char* extOut)` style "fill caller buffer" pour cohérence avec `read()`.

## Plan

1. ✅ Bridge gagne `sendCommand()` / `readResponse()` publics (commit `a12c090`).
2. ✅ Le flash compile contre la nouvelle API du bridge (commit `55eda25`), 65/65 tests natifs passent.
3. À toi : appliquer les conventions (`begin()`, préfixage des constants, README, etc.), rework des tests (UB + assertions vides), trancher le bug protocolaire `readSector` puis réactiver les 2 tests désactivés.
4. Câbler le smoke-test `src/main.cpp`.
5. On bouge la PR de draft à ready-for-review.

Pas de rush — la branche est rebasée, ton code compile, tu peux travailler dessus sans pression de conflit. Et pour les conventions, regarde aussi les retours pédagogiques sur #143, #152, #161 ; les patterns reviennent.

File: lib/daplink_flash/src/daplink_flash.cpp [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+#include "daplink_flash.h"
+
+#include <math.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <algorithm>
+#include <cctype>
+#include <cstdio>
+#include <cstring>
```
[the remark this answers is not in the document]
Comment (DumontALINE) [reply lib/daplink_flash/src/daplink_flash.cpp:11]: fix in commit : ac3a613
Add missing <vector> include to prevent build failures.

File: lib/daplink_flash/src/daplink_flash.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    // Set 8.3 filename. name: max 8 chars, ext: max 3 chars.
+    char n[DAPLINK_FLASH_FILENAME_LEN];
+    size_t nameLen = std::min(strlen(name), (size_t(DAPLINK_FLASH_FILENAME_LEN)));
+    for (int i = 0; i < nameLen; i++) {
+        n[i] = toupper((unsigned char)name[i]);
+    }
+
+    char e[DAPLINK_FLASH_EXT_LEN];
+    size_t extLen = std::min(strlen(ext), (size_t(DAPLINK_FLASH_EXT_LEN)));
+    for (int j = 0; j < extLen; j++) {
```
[the remark this answers is not in the document]
Comment (DumontALINE) [reply lib/daplink_flash/src/daplink_flash.cpp]: fix in commit : ac3a613
Add null checks in setFilename() to avoid null pointer crashes.

File: lib/daplink_flash/src/daplink_flash.cpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+DaplinkFlash::FilenameResult DaplinkFlash::getFilename() {
+    uint8_t raw[DAPLINK_FLASH_FILENAME_LEN + DAPLINK_FLASH_EXT_LEN];
+    _bridge->readResponse(DAPLINK_FLASH_CMD_GET_FILENAME, raw,
+                          DAPLINK_FLASH_FILENAME_LEN + DAPLINK_FLASH_EXT_LEN);
+    FilenameResult result;
+
+    memcpy(result.name, raw, DAPLINK_FLASH_FILENAME_LEN);
+    result.name[DAPLINK_FLASH_FILENAME_LEN] = '\0';
+    size_t nameLen = strnlen(result.name, DAPLINK_FLASH_FILENAME_LEN);
```
[the remark this answers is not in the document]
Comment (DumontALINE) [reply lib/daplink_flash/src/daplink_flash.cpp:75]: fix in commit : ac3a613
Zero-initialize raw buffer in getFilename() to prevent undefined behavior.

File: lib/daplink_flash/src/daplink_flash.cpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    // Erase entire flash memory.
+    _bridge->sendCommand(DAPLINK_FLASH_CMD_CLEAR_FLASH);
+}
+
+size_t DaplinkFlash::write(const uint8_t* data, size_t length) {
+    // Append data to the current file. Returns the number of bytes
+    if (data == nullptr && length == 0) {
+        return 0;
+    }
+
```
[the remark this answers is not in the document]
Comment (DumontALINE) [reply lib/daplink_flash/src/daplink_flash.cpp]: fix in commit : fdebad2
check data and length

File: lib/daplink_flash/src/daplink_flash.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        if (!_bridge->sendCommand(DAPLINK_FLASH_CMD_WRITE_DATA, payload, 1 + chunkLen)) {
+            return 0;
+        }
+        offset += chunkLen;
+    }
+    return length;
+}
+
+size_t DaplinkFlash::write(const char* data) {
```
[the remark this answers is not in the document]
Comment (DumontALINE) [reply lib/daplink_flash/src/daplink_flash.cpp:131]: fix in commit : fdebad2

File: lib/daplink_flash/README.md [hunk 6]
```diff
@@ -1,5 +1,81 @@
...
+| `size_t write(const uint8_t* data, size_t length)` | Append raw bytes to the current file. Returns the number of bytes written, or `0` on error. |
+| `size_t write(const char* data)` | Append a null-terminated string to the current file. |
+| `size_t writeLine(const char* text)` | Append a string followed by a newline character. |
+
+### Read operations
+
+| Method | Description |
+|--------|-------------|
+| `void readSector(uint16_t sector, uint8_t* buf)` | Read a 256-byte sector from flash into `buf`. |
+| `size_t read(uint8_t* result, size_t maxLen, bool limitLen = false)` | Read file content from flash. If `limitLen` is `false`, reads until the first `0xFF` sentinel. If `true`, reads up to `maxLen` bytes. Returns the number of bytes read. |
```
[the remark this answers is not in the document]
Comment (DumontALINE) [reply lib/daplink_flash/README.md]: fix in commit : fdebad2
Add missing readN() and readUntilSentinel() entries in README.

File: lib/daplink_flash/README.md [hunk 7]
```diff
@@ -1,5 +1,81 @@
...
+
+void setup() {
+    Serial.begin(115200);
+    internalI2C.begin();
+
+    flash.clearFlash();
+    flash.setFilename("DATA", "CSV");
+    flash.writeLine("temperature,pressure");
+    flash.writeLine("23.5,1013.2");
+}
```
[the remark this answers is not in the document]
Comment (DumontALINE) [reply lib/daplink_flash/README.md:38]: fix in commit : 9ded054
Add flash.begin() with error check before flash ops in README example.

File: lib/daplink_flash/library.properties [hunk 8]
```diff
@@ -6,3 +6,4 @@ sentence=Driver for the on-board DAPLink I2C-to-SPI flash bridge plus persistent
 category=Data Storage
 url=https://github.com/steamicc/arduino-steami-lib
 architectures=*
+depends=STeaMi DAPLink Bridge
```
[the remark this answers is not in the document]
Comment (DumontALINE) [reply lib/daplink_flash/library.properties:10]: fix in commit : 9ded054
Add includes=daplink_flash.h to library.properties for Arduino IDE.

File: tests/native/test_daplink_flash/test_main.cpp [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+int main(void) {
+    UNITY_BEGIN();
+    RUN_TEST(test_set_filename_sends_correct_payload);
+    RUN_TEST(test_get_filename_returns_stripped_name);
+    RUN_TEST(test_clear_flash_sends_cmd);
+    RUN_TEST(test_write_sends_data);
+    RUN_TEST(test_write_returns_length);
+    RUN_TEST(test_write_line_appends_newline);
+    RUN_TEST(test_read_sector_sends_correct_command);
+    RUN_TEST(test_read_stops_at_sentinel);
```
[the remark this answers is not in the document]
Comment (DumontALINE) [reply tests/native/test_daplink_flash/test_main.cpp:279]: fix in commit : 9ded054
Add RUN_TEST() for begin() tests and fix WHO_AM_I value in setUp().

**Comment by Charly-sketch** [comment]
Good work, I only have a few change requests before merge:

1. **`write(const char* data)` still needs a null guard.**
   This overload still calls `strlen(data)` directly, so `flash.write(nullptr)` would dereference a null pointer and crash. It should mirror the defensive check already added in `write(const uint8_t*, size_t)`.

2. **Read-path error handling is still too optimistic (`readSector()` / `getFilename()`).**
   `getFilename()` now zero-initializes its temporary buffer, which avoids UB, but it still ignores the number of bytes actually returned by `readResponse()`. Same issue in `readSector()`: there is no validation of `sendCommand()` success, no `buf == nullptr` guard, and no check that a full 256-byte sector was really received. Right now a bridge timeout or partial response silently turns into parsed garbage/default data instead of a detectable failure.

3. **`test_clear_flash_sends_cmd` is still not asserting anything meaningful.**
   The test computes `sawCmd` and then ends with `TEST_ASSERT_TRUE(true)`, so it passes unconditionally. Either assert `sawCmd`, or adjust the `TwoWire` mock if command-only frames are currently invisible there.

4. **Some protocol tests still validate too little.**
   A few tests only check that “a command happened” instead of checking the exact payload bytes:

   * `setFilename()` should assert uppercase + full space padding of the 8.3 payload.
   * `readSector(5)` should assert that the transmitted payload really contains the sector index bytes.
     Since this library is protocol glue, payload validation matters more than command presence alone.

5. **README still has a small API/doc mismatch.**
   In the Lifecycle table, `begin()` is documented as `void begin()` while the real implementation is `bool begin()`. There is also a malformed Markdown row before `readUntilSentinel()`.

Once those are cleaned up, I’m happy with the direction of the PR.

File: tests/hardware/test_daplink_flash/test_main.cpp [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    for (size_t i = 0; i < strlen(current.name); ++i) {
+        unsigned char c = static_cast<unsigned char>(current.name[i]);
+        TEST_ASSERT_TRUE_MESSAGE(c >= 0x20 && c < 0x7F, "filename contains non-printable byte");
+    }
+    for (size_t i = 0; i < strlen(current.ext); ++i) {
+        unsigned char c = static_cast<unsigned char>(current.ext[i]);
+        TEST_ASSERT_TRUE_MESSAGE(c >= 0x20 && c < 0x7F, "extension contains non-printable byte");
+    }
+}
```
[the remark this answers is not in the document]
Comment (nedseb) [reply tests/hardware/test_daplink_flash/test_main.cpp:56]: fix in 9c6cc64 — setUp() now asserts flash.begin() so a missing bridge fails fast instead of letting empty filename strings make the printable-ASCII test pass vacuously.

File: lib/daplink_bridge/src/DaplinkBridge.cpp [hunk 11]
```diff
@@ -122,6 +122,78 @@ size_t DaplinkBridge::readConfig(uint8_t* result, size_t maxLen) {
...
+    while (produced < maxLen) {
+        if (!waitNotBusy(timeoutMs)) {
+            return produced;
+        }
+
+        const uint8_t want = static_cast<uint8_t>(
+            std::min<size_t>(DAPLINK_BRIDGE_MAX_READ_CHUNK, maxLen - produced));
+
+        readBlock(static_cast<uint8_t>(DAPLINK_BRIDGE_REG_RESPONSE + produced), buf + produced,
+                  want);
```
[the remark this answers is not in the document]
Comment (nedseb) [reply lib/daplink_bridge/src/DaplinkBridge.cpp]: fix in 8b3b27f — readResponse now emits the cmd byte once and streams the response through successive requestFrom() calls without ever recomputing REG_RESPONSE+produced, so the 0x82+0x7D=0xFF wraparound disappears. The chunk loop also no longer calls waitNotBusy()/readBlock() between chunks, which were the two things that reset the firmware response cursor.

File: lib/daplink_flash/library.properties [hunk 12]
```diff
@@ -6,3 +6,5 @@ sentence=Driver for the on-board DAPLink I2C-to-SPI flash bridge plus persistent
 category=Data Storage
 url=https://github.com/steamicc/arduino-steami-lib
 architectures=*
+includes=daplink_flash.h
```
[the remark this answers is not in the document]
Comment (nedseb) [reply lib/daplink_flash/library.properties:9]: Acknowledged — deferring to a follow-up PR/issue to keep this one focused on the review feedback. Tracking the rename of daplink_flash.{h,cpp,_const.h} (and the matching library.properties / README / #include sites) separately so it can be reviewed on its own.

File: tests/native/helpers/Wire.h [hunk 13]
```diff
@@ -35,7 +35,9 @@ class TwoWire {
             }
             currentRegisterByAddr_[currentAddress_] = reg;
         } else if (txBuffer_.size() == 1) {
-            currentRegisterByAddr_[currentAddress_] = txBuffer_[0];
+            uint8_t cmd = txBuffer_[0];
+            commands_.push_back({currentAddress_, cmd});
+            currentRegisterByAddr_[currentAddress_] = cmd;
         }
```
[the remark this answers is not in the document]
Comment (nedseb) [reply tests/native/helpers/Wire.h:56]: fix in 9c6cc64 — requestFrom() now pops the trailing entry from commands_ when it matches the current address, so a register-pointer-select single-byte transmission followed by a read no longer pollutes getCommands() with a phantom command.

File: lib/daplink_flash/README.md [hunk 14]
```diff
@@ -1,5 +1,86 @@
...
+    if (!flash.begin()) {
+        Serial.println("DAPLink Flash not found!");
+        return;
+    }
+
+    flash.clearFlash();
+    flash.setFilename("DATA", "CSV");
+    flash.writeLine("temperature,pressure");
+    flash.writeLine("23.5,1013.2");
+}
```
[the remark this answers is not in the document]
Comment (nedseb) [reply lib/daplink_flash/README.md:38]: fix in 9c6cc64 — README quick start now opens with a callout warning that clearFlash() wipes the whole partition, and the snippet itself carries an inline comment flagging the destructive call.

File: tests/native/test_daplink_flash/test_main.cpp [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+
+    if (foundCmd) {
+        const auto& writes = Wire.getWrites();
+        if (writes.size() >= 10) {
+            payloadCorrect =
+                (writes[0].value == 'T') && (writes[1].value == 'E') && (writes[2].value == 'S') &&
+                (writes[3].value == 'T') && (writes[4].value == ' ') && (writes[5].value == ' ') &&
+                (writes[6].value == ' ') && (writes[7].value == ' ') && (writes[8].value == 'T') &&
+                (writes[9].value == 'X') && (writes[10].value == 'T');
+        }
```
[the remark this answers is not in the document]
Comment (nedseb) [reply tests/native/test_daplink_flash/test_main.cpp:63]: fix in 9c6cc64 — bumped the size check from >= 10 to >= 11 so the writes[10] read is bounds-safe.

File: lib/daplink_flash/src/daplink_flash.cpp [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+        e[j] = toupper((unsigned char)ext[j]);
+    }
+
+    uint8_t padded[DAPLINK_FLASH_FILENAME_LEN + DAPLINK_FLASH_EXT_LEN];
+    memset(padded, ' ', sizeof(padded));
+    memcpy(padded, n, nameLen);
+    memcpy(padded + DAPLINK_FLASH_FILENAME_LEN, e, extLen);
+
+    _bridge->sendCommand(DAPLINK_FLASH_CMD_SET_FILENAME, padded, sizeof(padded));
+}
```
[the remark this answers is not in the document]
Comment (nedseb) [reply lib/daplink_flash/src/daplink_flash.cpp]: fix in 9c6cc64 — setFilename() now returns bool, propagating null-pointer rejection and the bridge sendCommand() result. PR body API box updated to match.
